src/auto-evo/mutation_strategy/UpgradeOrganelle.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using AutoEvo;
+
+internal class UpgradeOrganelle : IMutationStrategy<MicrobeSpecies>
+{
+    private readonly FrozenSet<OrganelleDefinition> allOrganelles;
+    private IComponentSpecificUpgrades upgrade;
+
+    public UpgradeOrganelle(Func<OrganelleDefinition, bool> criteria, IComponentSpecificUpgrades upgrade)
+    {
+        allOrganelles = SimulationParameters.Instance.GetAllOrganelles().Where(criteria).ToFrozenSet();
+        this.upgrade = upgrade;
+    }
+
+    public bool Repeatable => true;
+
+    public List<Tuple<MicrobeSpecies, float>>? MutationsOf(MicrobeSpecies baseSpecies, float mp, bool lawk,
+        Random random)
+    {
+        if (allOrganelles.Count == 0)
+        {
+            return null;
+        }
+
+        var mutated = new List<Tuple<MicrobeSpecies, float>>();
+
+        MicrobeSpecies newSpecies = (MicrobeSpecies)baseSpecies.Clone();
```

Comment:
This still has the problem where this clones a species before knowing that it would actually apply any upgrades. So I'd prefer an approach where a copy is only made *after* it is known that it is possible. Also allocating the `mutated` list once needed would be a good memory use optimization.